http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java

```diff
@@ -1142,110 +1193,167 @@ protected void onComplete(MutableHttpResponse<?> message) {
+                                                    finalResponse,
+                                                    result,
+                                                    defaultResponseMediaType
+                                            );
                                         }
-                                        encodeHttpResponse(
-                                                context,
-                                                request,
-                                                finalResponse,
-                                                o,
-                                                defaultResponseMediaType
-                                        );
                                     }
                                 });
-
                             } else {
                                 // streaming case
                                 Argument<?> typeArgument = finalRoute.getReturnType().getFirstTypeVariable().orElse(Argument.OBJECT_ARGUMENT);
                                 boolean isHttp2 = request.getHttpVersion() == io.micronaut.http.HttpVersion.HTTP_2_0;
                                 if (HttpResponse.class.isAssignableFrom(typeArgument.getType()) && !typeArgument.getFirstTypeVariable().map(Argument::isAsyncOrReactive).orElse(false)) {
                                     // a response stream
-                                    Flowable<HttpResponse<?>> bodyFlowable = Publishers.convertPublisher(body, Flowable.class);
-
+                                    Publisher<HttpResponse<?>> bodyPublisher = Publishers.convertPublisher(body, Publisher.class);
                                     // HTTP/2 allows sending multiple responses down a single stream
                                     if (isHttp2) {
-                                        bodyFlowable.subscribe(httpResponse -> encodeHttpResponse(
-                                                context,
-                                                request,
-                                                toNettyResponse(httpResponse),
-                                                httpResponse.body(),
-                                                defaultResponseMediaType
-                                        ), throwable -> exceptionCaughtInternal(
-                                                context,
-                                                throwable,
-                                                request,
-                                                false
-                                        ));
+                                        bodyPublisher.subscribe(new CompletionAwareSubscriber<HttpResponse<?>>() {
+                                            @Override
+                                            protected void doOnSubscribe(Subscription subscription) {
+                                                subscription.request(Long.MAX_VALUE);
```

Comment:
Do we not need to control back pressure here requesting 1 writing the message and then requesting the next?

Reply:
Original code didn't do it, but I think it makes sense to have it.